### PR TITLE
support multi-primary key

### PR DESCRIPTION
**Pull request by andyqian**
### Ⅰ. Describe what this PR did

1. support multi-primary key

### CODE REVIEW

**Comment by xingfudeshi** [comment]
> ### Ⅰ. Describe what this PR did
> 1. support multi-primary key

CI is failed.

**Comment by andyqian** [comment]
> > ### Ⅰ. Describe what this PR did
> > 
> > 1. support multi-primary key
> 
> CI is failed.

done

**Comment by ujjboy** [comment]
@andyqian Thanks for your pr. can you add more test case ?  such as adding the case of the multi-primary key in `BaseH2Test.java`

**Comment by andyqian** [comment]
ok


 
---Original---

**Comment by zhangthen** [comment]
@andyqian   please fix the CI  problems.

**Comment by andyqian** [comment]
> @andyqian please fix the CI problems.

done

File: rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java [hunk 1]
```diff
@@ -27,9 +27,7 @@
 import java.math.BigDecimal;
 import java.sql.Timestamp;
 import java.sql.Types;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java]: forbid to import *

**Comment by andyqian** [comment]
ok，fix it later


 
---Original---

File: rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java (hunk 1 above)
Comment (andyqian) [reply rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java]: > forbid to import *

done

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java [hunk 2]
```diff
@@ -127,12 +127,11 @@ public void add(Row row) {
      * @return the list
      */
     public List<Field> pkRows() {
-        final String pkName = getTableMeta().getPkName();
         List<Field> pkRows = new ArrayList<>();
         for (Row row : rows) {
             List<Field> fields = row.getFields();
             for (Field field : fields) {
-                if (field.getName().equalsIgnoreCase(pkName)) {
+                if (getTableMeta().containsPK(field.getName())) {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java]: Does your code really need to be modified here? ref #1218

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java [hunk 3]
```diff
@@ -180,7 +180,7 @@ public static TableRecords buildRecords(TableMeta tmeta, ResultSet resultSet) th
                 ColumnMeta col = tmeta.getColumnMeta(colName);
                 Field field = new Field();
                 field.setName(col.getColumnName());
-                if (tmeta.getPkName().equalsIgnoreCase(field.getName())) {
+                if (tmeta.containsPK(field.getName())) {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java]: same above

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java [hunk 4]
```diff
@@ -115,6 +114,17 @@ protected boolean containsPK() {
...
 
 
+    /**
+     * check columnName is primary key
+     * @param columnName columnName
+     * @return true or false
+     */
+    private boolean containsPkName(String columnName) {
+        List<String> primaryKeyRowNames = getTableMeta().getPrimaryKeyOnlyName();
+        return primaryKeyRowNames.contains(columnName);
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: equalsIgnoreCase

**Comment by andyqian** [comment]
ok


 
---Original---

File: rm-datasource/src/test/java/io/seata/rm/datasource/undo/UndoExecutorTest.java [hunk 5]
```diff
@@ -43,9 +43,7 @@
 import java.sql.Time;
 import java.sql.Timestamp;
 import java.sql.Types;
-import java.util.Calendar;
-import java.util.Map;
-import java.util.Properties;
+import java.util.*;
```
Comment (slievrly) [inline rm-datasource/src/test/java/io/seata/rm/datasource/undo/UndoExecutorTest.java]: class count to use import * -> 999

**Comment by slievrly** [comment]
> ok
ref #1218

File: rm-datasource/src/test/java/io/seata/rm/datasource/undo/UndoExecutorTest.java (hunk 5 above)
Comment (andyqian) [reply rm-datasource/src/test/java/io/seata/rm/datasource/undo/UndoExecutorTest.java]: done

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java (hunk 4 above)
Comment (andyqian) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/InsertExecutor.java]: done

**Comment by ujjboy** [comment]
hi @andyqian  please resolve conflicts.

File: rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java [hunk 6]
```diff
@@ -154,23 +156,40 @@ private static boolean compareRows(TableMeta tableMetaData, List<Row> oldRows, L
...
+    /**
+     * get RowKey
+     * @return rowKey
+     */
+    private static String getRowKey(Set<String> fields, Map<String,Object> rowsKeysMap){
+        List<String> list = new ArrayList<>(fields);
+        Collections.sort(list);
+        StringBuilder rowKey = new StringBuilder();
+        for(String item:list){
+            rowKey.append(rowsKeysMap.get(item));
```
[the remark this answers is not in the document]
Comment (andyqian) [reply rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java]: good idea

File: rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java [hunk 7]
```diff
@@ -154,23 +156,40 @@ private static boolean compareRows(TableMeta tableMetaData, List<Row> oldRows, L
...
         return rowMap;
     }
 
+    /**
+     * get RowKey
+     * @return rowKey
+     */
+    private static String getRowKey(Set<String> fields, Map<String,Object> rowsKeysMap){
+        List<String> list = new ArrayList<>(fields);
+        Collections.sort(list);
```
[the remark this answers is not in the document]
Comment (andyqian) [reply rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java]: I thought about it，How to get it ?

**Comment by andyqian** [comment]
> hi @andyqian please resolve conflicts.

done

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableMetaCache.java [hunk 8]
```diff
@@ -97,15 +90,15 @@ public static TableMeta getTableMeta(final DataSourceProxy dataSourceProxy, fina
...
+    public static void refresh(final DataSourceProxy dataSourceProxy) {
         ConcurrentMap<String, TableMeta> tableMetaMap = TABLE_META_CACHE.asMap();
         for (Entry<String, TableMeta> entry : tableMetaMap.entrySet()) {
             try {
                 TableMeta tableMeta = fetchSchema(dataSourceProxy, entry.getValue().getTableName());
                 if (tableMeta == null){
                     LOGGER.error("get table meta error");
                 }
-                if (!tableMeta.equals(entry.getValue())){
+                if (!tableMeta.equals(entry.getValue())) {
```
Comment (jsbxyyx) [inline rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableMetaCache.java]: tableMeta may be appears NPE

**Comment by l81893521** [comment]
Hi. Multi pk already done at #2398 , so, I will close this pr. Hope for you next contribution.

**Pull request closed without merging**